### PR TITLE
JM - 4F - Button and Modal to Remove and Restore Users (Merge after B3, 24)

### LINKED ISSUE #26: #4 - F1 - Button to suspend/restore users to Admin/User Table
Just add the information to the table about whether user is supended or not ,and the button that either says "Suspend" or "Restore". The buttons should work, but the confirmation modal can be saved for a second issue.

### LINKED ISSUE #27: #4 - F2 - Confirmation modal to button that suspends or restores users
Add a confirmation modal to the button that suspends or restores a user.

### PR DESCRIPTION
**Pull request by Jumshim**
## Overview
This PR builds off of `4-B3-JeffreyRemoveUsers` and touches 5 different files: `usersFixtures.js`, `UsersTable.js`, `users.js`, and their relevant test files. This PR also address both issue #26 -- Button to suspend/restore users -- and issue #27  -- Confirmation modal to suspend/restore users -- in the same PR.

**NOTE**: The prior PR for `4-B3-JeffreyRemoveUsers` and the PR for `24-JeffreyIncrementalPitest` must be merged before touching this one.

Link to dokku deployment is below in the Validation section.
### Workflow to Suspend
image

image

image

### Workflow to Restore
image

image

image

## Feedback Request (Optional)
- I colored the `Restore` and `Suspend` buttons differently to differentiate them. Is this a welcome coloring?
- Additionally, I chose to use the user's email rather than their name for identification purposes because that's been the standard for the backend as well (user can only be queried by email).

## Validation
1. Navigate to the dokku deployment: https://happycows-qa.dokku-08.cs.ucsb.edu 
2. Log In
3. Go to this page -- https://happycows-qa.dokku-08.cs.ucsb.edu/admin/users
4. Test this function.
5. Refresh to see changes

## Tests
- [X] Frontend Unit tests (`npm test`) pass
- [X] Frontend Test coverage (`npm run coverage`) 100%
- [X] Frontend Mutation tests (`npx stryker run`) 100% 
- [X] Frontend Linting (`npx eslint --fix src`) 

## Linked Issues

Closes #26 
Closes #27

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,46 +2,43 @@
```
Comment (Jumshim) [inline README.md]: All of these changes to the README.md are from my `Prettier` extension; can be ignored.

File: frontend/src/fixtures/usersFixtures.js [hunk 2]
```diff
@@ -1,48 +1,54 @@
 const usersFixtures = {
```
Comment (Jumshim) [inline frontend/src/fixtures/usersFixtures.js]: For fixtures, I simply added the `suspended` field to each of our examples for consistency.

**Review by niyati-m — DISMISSED** [review]
Looks great to me!

File: README.md [hunk 3]
```diff
@@ -123,3 +119,25 @@ In order to run the end-to-end tests 'not headless' use the following instead of
...
+For example, to run pitest on just one class, use:
+
+```
+mvn pitest:mutationCoverage -DtargetClasses=edu.ucsb.cs156.example.controllers.RestaurantsController
+```
+
+To run pitest on just one package, use:
+
+```
+mvn pitest:mutationCoverage -DtargetClasses=edu.ucsb.cs156.example.controllers.*
```
Comment (pconrad) [inline README.md]:
```suggestion
mvn pitest:mutationCoverage -DtargetClasses=edu.ucsb.cs156.example.controllers.\*
```

Need to fix it in this branch too or we'll get merge conflicts.  This is why it's best to keep separate independent branches where possible.

File: README.md (hunk 1 above)
Comment (pconrad) [reply README.md]: It's really not a great idea to include changes like this in an unrelated PR.  It makes code review and merging much harder.  We can't just "ignore" them, because in this case, there was a typo introduced in another branch that this would persist.

And it's likely to create merge conflicts.  I want to encourage you to restore this file to the way it is on the main branch, and if you want to update the formatting, do it in a separate PR.

File: frontend/src/main/utils/users.js [hunk 4]
```diff
@@ -2,16 +2,60 @@ import { useQuery } from "react-query";
...
-  return useQuery("users", async () => {
-    const uri = "/api/admin/users";
-    try {
-      const response = await axios.get(uri);      
-      return response.data ;
-    } catch (e) {
-      console.error(`Error getting data from ${uri}:`,e);
-      return [];
+  return useQuery(
+    "users",
```
Comment (pconrad) [inline frontend/src/main/utils/users.js]: So the query key here is "users".  It's going to be important that we invalidate that key as soon as you suspend or restore a user so that the table is updated.

File: frontend/src/main/utils/users.js [hunk 5]
```diff
@@ -2,16 +2,60 @@ import { useQuery } from "react-query";
...
+        params: {
+          userId: user.id,
+        },
+      }
+    );
+    return response.data;
+  } catch (e) {
+    console.error(`Error restoring user with id ${user.id} from ${uri}`, e);
+    return [];
+  }
```
Comment (pconrad) [inline frontend/src/main/utils/users.js]: Same here.  This should be a useBackendMutation or at least something that invalidates the react query key for "users" so that the users table refreshes.

File: pom.xml [hunk 6]
```diff
@@ -262,7 +263,7 @@
                     </historyOutputFile>
                     <verbose>true</verbose>
                     <targetClasses>
-                        <param>edu.ucsb.cs156.*</param>
+                        <param>${targetClasses}</param>
                     </targetClasses>
                     <targetTests>
                         <param>edu.ucsb.cs156.*</param>
```
Comment (pconrad) [inline pom.xml]: This should be in a separate PR.  Don't mix concerns that have nothing to do with each other in the same PR.

See: https://ucsb-cs156.github.io/topics/git/git_feature_branch_workflow.html#more-detail-about-creating-feature-branches

**Review by pconrad — CHANGES_REQUESTED** [review]
Please review the PR feedback. There are a few important issues here.

File: README.md (hunk 1 above)
Comment (Jumshim) [reply README.md]: I've reverted the README back to the main branch! Everything should be cleared now :)

Also, yes I agree. I just wanted to use my individual pitest coverage stuff on this branch xD

File: frontend/src/main/utils/users.js (hunk 5 above)
Comment (Jumshim) [reply frontend/src/main/utils/users.js]: Done!

File: pom.xml (hunk 6 above)
Comment (Jumshim) [reply pom.xml]: Fixed!

File: frontend/src/main/utils/users.js (hunk 4 above)
Comment (Jumshim) [reply frontend/src/main/utils/users.js]: Done!

File: frontend/src/main/utils/users.js [hunk 7]
```diff
@@ -1,17 +1,48 @@
...
+
+export function useSuspendUser() {
+  return useBackendMutation(
+    (cell) => ({
+      method: "post",
+      url: "/api/admin/users/suspend",
+      params: { userId: cell.row.values.id },
+    }),
+    {},
+    // Stryker disable next-line all : hard to set up test for query key invalidation for mutations
```
Comment (Jumshim) [inline frontend/src/main/utils/users.js]: When I Ctrl+F'd the codebase for other examples that used useBackendMutation, I noticed they all disable checking for the "users" query key invalidation because it was hard to simulate / test.
